tools/autograd/derivatives.yaml

```diff
@@ -789,7 +789,14 @@
 - name: scatter_add_(Tensor(a!) self, int dim, Tensor index, Tensor src) -> Tensor(a!)
   self: grad
   index: non_differentiable
-  src: grad.gather(dim, index)
+# previously it used to be: grad.gather(dim, index),
```

Comment:
Why not `  src: grad.gather(dim, index).expand(src.sizes())` ?
Or `  src: grad.gather(dim, index.expand(src.sizes()))` to match the implementation you currently have.

Reply:
The first option seems reasonable, the second one, with the index expanded, is not correct. I will check how expand works...

Reply:
Ho right, "broadcastable" and "expandable" are not the same thing?

Reply:
This is what I get with expand:
```python
In [3]: a = torch.randn(3, 3)

In [4]: a
Out[4]: 
tensor([[ 0.1124, -1.3161,  0.5052],
        [ 0.0658,  0.9937, -1.4280],
        [ 0.3332,  0.1201,  0.1397]])

In [5]: a.expand(5, 5)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-d54bdcaa6cfb> in <module>
----> 1 a.expand(5, 5)

RuntimeError: The expanded size of the tensor (5) must match the existing size (3) at non-singleton dimension 1.  Target sizes: [5, 5].  Tensor sizes: [3, 3]

```

Reply:
Ho wait I think I'm missing something here then :D
The issue you linked mentions the case where the indices are broadcastable to the src. And so are broadcasted. They mention that expanding the indices before fixes their issue.

Do you consider a different case here?

Reply:
No, broadcasting is not relevant here :) Scatter_add samples a subset of src, while gather, which is used in backward of scatter_add, creates an output which of the same shape as index, but it should be of the same shape as src...

Reply:
> Scatter_add samples a subset of src

That's where my problem is :D 
In my mind, you should have one index of every source element that tells you where to scatter it.
Is it not the case?

Reply:
Yes, this is the case. But, the issue is that for scatter_add you have src.shape >= index.shape. But gather always outputs something of the same shape as index, but it should match src.shape. Hence, we create a zero tensor of shape src.shape, and gather into it. Maybe we could use gather_out, actually...

Reply:
Ho right. I remember. That does sound like a bug in the shape checking to me... But I guess it is documented for scatter add so its fine.

Now for gather. The output should have the right size.
But gather_stub will only use the beginning of each dimension properly and leave the rest as 0 in the case where the provided output is too big? Is that the expected behavior here? Shall we mention in a comment that we use this behavior that is not part of the original gather function?

Reply:
Yes, it has to be zero, because no gradient goes there. This is by plan. That is why I decided to make a separate backward method and not like a separate gather-like function.

Reply:
Besides, the internal gather_stub checks make sure that the number of dimensions of result and index are the same.

Reply:
Ok sounds good.

Reply:
There is a feature request for a function that could be used here, actually.
[Like this](https://github.com/pytorch/pytorch/issues/31913).